pgjdbc/src/main/java/org/postgresql/core/v3/SimpleQuery.java

```diff
@@ -114,11 +118,29 @@ void setStatementName(String statementName, short deallocateEpoch) {
     this.deallocateEpoch = deallocateEpoch;
   }
 
-  void setStatementTypes(int[] paramTypes) {
-    this.preparedTypes = paramTypes;
+  void setPrepareTypes(int[] paramTypes) {
+    // Remember which parameters were unspecified since the parameters will be overriden later by
+    // ParameterDescription message
+    for (int i = 0; i < paramTypes.length; i++) {
+      int paramType = paramTypes[i];
+      if (paramType == Oid.UNSPECIFIED) {
+        if (this.unspecifiedParams == null) {
+          this.unspecifiedParams = new BitSet();
+        }
+        this.unspecifiedParams.set(i);
+      }
+    }
+
+    // paramTypes is changed by "describe statement" response, so we clone the array
+    // However, we can reuse array if there is one
+    if (this.preparedTypes == null) {
+      this.preparedTypes = paramTypes.clone();
+      return;
+    }
+    System.arraycopy(paramTypes, 0, this.preparedTypes, 0, paramTypes.length);
   }
 
-  int[] getStatementTypes() {
+  int[] getPrepareTypes() {
     return preparedTypes;
```

Comment:
Should this return a clone to prevent mutation?

Reply:
This is a private API, and there's just one usage. No need to clone.